README.rst
MSSQL Runner
============

.. image:: https://badge.fury.io/py/mssql-runner.svg
    :target: https://badge.fury.io/py/mssql-runner
    :alt: PyPI Version

.. image:: https://readthedocs.org/projects/mssql-runner/badge/?version=latest
    :target: https://mssql-runner.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://img.shields.io/badge/Contributor%20Covenant-v2.0%20adopted-ff69b4.svg
    :target: https://github.com/equinoxfitness/datacoco-email_tools/blob/master/CODE_OF_CONDUCT.rst
    :alt: Code of Conduct

MSSQL Runner provides a way of running MSSQL script with a set of parameters for ETL usage

Installation
------------

MSSQL Runner requires Python 3.6+

::

    python3 -m venv <virtual env name>
    source <virtual env name>/bin/activate
    pip install mssql-runner

Quickstart
----------

There are three types of substitution parameters that can be used through mssql runner.  The first is a set of standard etl params for ease of use:

*  -f, from_date, default 1776-07-24
*  -t,to_date, default 9999-12-31
*  -b, batch_no, default is -1

The second option is for arbitrary variable expansion.  This is passed in the following format because rundeck commands don't like json quotes:

*  -p, example: -p "param1-val1, param2-val2"

The final option is to use the batchy integration

*  -wf, batchy_job, this will substitue parameters from a batchy workflow, this should be a fully qualified batchy job name of the format wf.job, if no job is specified it will assume global

You also have the option to choose which config provider to use.

*  -conf, config, default core

Datacoco Core - This is default config provider which is looking for a file called etl.cfg in your project root.

etl.cfg folder structure

```
[sample]
db_name=local
user=user
server=server
password=XXXX
port=1433
type=mssql
```

Secret Manager - Using this option assumes that you already have secret manager setup in your aws account and as aws key and secret is configured in your environment for you to connect to aws.

Here's example how to use secret manager config option
*  -conf secret_manager --secret_project_name <your_project> --secret_team <your_team>


Here is a sample SQL Script.  If run in SQL workbench you will be prompted for values for var1 and var2.

::

    drop table if exists  zzztemp;

    create table zzztemp (
    dt timestamp,
    var varchar,
    from_date timestamp,
    batch_no integer
    );

    insert into zzztemp
    values (getdate(), '$[?var1]', '$[?from_date]', '$[?batch_no]');

    insert into zzztemp
    values (getdate(), '$[?var2]', '$[?from_date]', '$[?batch_no]');

    select * from zzztemp;


In mssql runner you would use the following params to substitute that value.  It is assumed these params would be dynamically substituted by the calling script or informatica process:

``python -m mssql_runner.module -s "sample/mssql_runner_test.sql" -p "var1-cat, var2-dog" -b '9999'``

Assuming you had workflow config in batchy under wf3, you could also use this script:

``python -m  mssql_runner.module -s sample/mssql_runner_test.sql -wf wf3``


Development
-----------

Getting Started
~~~~~~~~~~~~~~~

It is recommended to use the steps below to set up a virtual environment for development:

::

    python3 -m venv <virtual env name>
    source <virtual env name>/bin/activate
    pip install -r requirements.txt

Testing
~~~~~~~

::

    pip install -r requirements-dev.txt

To run the testing suite, simply run the command: ``tox`` or ``python -m unittest discover tests``

Contributing
------------

Contributions to mssql\_runner are welcome!

Please reference guidelines to help with setting up your development
environment
`here <https://github.com/equinoxfitness/mssql-runner/blob/master/CONTRIBUTING.rst>`__.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.6] - 2020-04-07
### Changed
    - Change database config key from server to host

## [0.1.5] - 2020-03-20
### Changed
    - Fix issue when getting Batch config

## [0.1.4] - 2020-03-19
### Changed
    - Secret manager default project name

## [0.1.3] - 2020-03-18
### Changed
    - Added documentation how to use config manager
    - Secret manager config parameter

## [0.1.2] - 2020-03-17
### Added
    - Add default from date
### Changed
    - Update setup.py missing dependency packages

## [0.1.1] - 2020-03-17
### Added
    - mssql-runner for cli execution
    - Update tests
    - nose.cfg for tox
    - update requirements.txt dependencies

### Changed
    - Modify mssql_runner to make it executable as script and as module call

## [0.1.0] - 2020-03-11
### Added
    - Wrapper config class
    - MSSql runner classes
    - Batch control

[0.1.6]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.5...v0.1.6
[0.1.5]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.4...v0.1.5
[0.1.4]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/equinoxfitness/mssql-runner/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/equinoxfitness/mssql-runner/releases/tag/v0.1.0


tests/test_mssql_runner.py
import os
import unittest
from unittest.mock import MagicMock
from mssql_runner import UNIT_TEST_KEY
from mssql_runner.module import MSSQLRunner


class TestMSSQLRunner(unittest.TestCase):
    def setUp(self):
        os.environ[UNIT_TEST_KEY] = 'True'
        self.testCls = MSSQLRunner(
            db_name='', host='', user='', password='', port='', batchy_server='', batchy_port=''
        )

    def test_run_script(self):
        self.testCls.ms = MagicMock()
        self.testCls.run_script(script='',
                                params=None,
                                sql_command='select * from test_db')

    def test_run_script_with_params(self):
        self.testCls.ms = MagicMock()
        self.testCls.run_script(script='',
                                params='title-test',
                                sql_command='''
                                    select * from test_db
                                    where title=$[?title]
                                ''')

    def test_run_script_with_params_multiple(self):
        self.testCls.ms = MagicMock()
        self.testCls.run_script(script='',
                                params='title-test,firstname-john,lastname-doe',
                                sql_command='''
                                    select * from test_db
                                    where title=$[?title]
                                    and firstname=$[?firstname]
                                    and lastname=$[?lastname]
                                ''')

    def test_run_script_from_date(self):
        self.testCls.ms = MagicMock()
        self.testCls.run_script(script='',
                                params=None,
                                from_date='2020-03-10T0:00:00.00',
                                sql_command='''
                                    select * from test_db
                                    where from_date=$[?from_date]
                                ''')

    def test_run_script_others(self):
        self.testCls.ms = MagicMock()
        self.testCls.run_script(script='',
                                params=None,
                                from_date='2020-03-10T0:00:00.00',
                                to_date='2020-03-12',
                                batch_id='1',
                                sql_command='''
                                    select * from test_db
                                    where from_date=$[?from_date]
                                ''')


mssql_runner/module.py
#!/usr/bin/env python

import sys
import os
import argparse
from datacoco_core.logger import Logger
from datacoco_batch import Batch
from datacoco_db.mssql_tools import MSSQLInteraction
from datetime import datetime
from mssql_runner import DEFAULT_FROM_DATE

from mssql_runner import UNIT_TEST_KEY
from mssql_runner.config_wrapper import ConfigWrapper


class MSSQLRunner:
    """
    generic class for execution of a parameterized script
    in postgres or redshift
    """

    def __init__(self, db_name, host, user, password, port, batchy_server, batchy_port):
        self.ms = None

        self.is_test = os.environ.get(UNIT_TEST_KEY, False)

        self.batchy_server = batchy_server
        self.batchy_port = batchy_port

        if not self.is_test:
            self.ms = MSSQLInteraction(
                dbname=db_name,
                host=host,
                user=user,
                password=password,
                port=port,
            )
            self.ms.conn()

    @staticmethod
    def expand_params(sql, params):
        """
        substitutes params in sql stagement
        :param sql:
        :param params:
        :return: sql, expanded with params
        """
        for p in params.keys():
            var = "$[?" + p + "]"
            val = str(params[p])
            if p == "from_date" and val != "1776-07-04":
                try:
                    val = datetime.strptime(val, "%Y-%m-%dT%H:%M:%S.%f")
                    val = val.strftime("%Y-%m-%d %H:%M:%S")
                    print("removed microsecond from wf from_date")
                except Exception as e:
                    print(e)
                    pass
            sql = sql.replace(var, val)
        return sql

    def run_script(
        self,
        script,
        from_date=None,
        to_date=None,
        batch_id=None,
        params=None,
        batchy_job=None,
        sql_command=None,
    ):
        """
        method for expanding and running sql statements
        :param script:
        :param from_date:
        :param to_date:
        :param batch_id:
        :param params:
        :return:
        """

        paramset = {}

        # set up logger
        if script is not None:
            script_path, script_filename = os.path.split(script)
            logger = Logger(logname=script_filename)
        else:
            logger = Logger(logname="direct_sql")
            sys.excepthook = logger.handle_exception

        # first we retrive params  we will load these into dict first,
        # any additional params specified will override
        if batchy_job:
            wf = batchy_job.split(".")[0]
            try:
                job = (
                    batchy_job.split(".")[1]
                    if len(batchy_job.split(".")[1]) > 0
                    else "global"
                )
            except Exception as e:
                logger.l(e)
                job = "global"
            batchy_params = Batch(
                wf,
                server=self.batchy_server,
                port=self.batchy_port,
            ).get_status()
            paramset.update(batchy_params[job])

        # next we apply custom params and special metadata fields,
        # again this will overrite batchy params if specified
        # convert string params to dict
        try:
            params = dict(
                (k.strip(), v.strip())
                for k, v in (item.split("-") for item in params.split(","))
            )
        except Exception as e:
            logger.l(f"issue parsing params: {e}")

        if isinstance(params, dict):
            paramset.update(params)

        if from_date:
            paramset["from_date"] = from_date
        if to_date:
            paramset["to_date"] = to_date
        if batch_id:
            paramset["batch_id"] = batch_id

        # now defaults for special metadata fields
        if paramset.get("from_date") is None:
            paramset["from_date"] = DEFAULT_FROM_DATE
        if paramset.get("to_date") is None:
            paramset["to_date"] = "9999-12-31"
        if paramset.get("batch_id") is None:
            paramset["batch_id"] = "-1"
        # we'll keep batch_no for backwards compatibility
        paramset["batch_no"] = paramset["batch_id"]

        # check if it's direct SQL or SQL file
        if not self.is_test and script is not None:
            try:
                raw_sql = open(script).read()
            except Exception as e:
                e = f"File not found, please check file path."
                logger.l(e)
                raise RuntimeError(e)
        else:
            raw_sql = sql_command
        sql = self.expand_params(raw_sql, paramset)
        sql_message = (
            "\n\n--sql script start:\n" + sql + "\n--sql script end\n\n"
        )
        logger.l(sql_message)

        self.ms.batch_open()

        logger.l("starting script")
        try:
            self.ms.exec_sql(sql)
            self.ms.batch_commit()
            logger.l("batch commit")
        except Exception as e:
            logger.l("execution failed with error: " + str(e))
            raise RuntimeError(e)
            # os._exit(1)  # this truncates logout in rundeck


def main(args):

    # argparse
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-s", "--script", help="""enter a path to your script """, default=None
    )
    # rundeck doesn't like -p '{"param1":"val1", "param2":"val2"}',
    # changed params to string
    parser.add_argument(
        "-p",
        "--parameters",  # type=json.loads, default='{"none":"none"}',
        default="none-none",
        help="""additional params to be substituted in script,
        example: -p param1-val1, param2-val2 """,
    )
    parser.add_argument(
        "-d",
        "--database",
        help="""db alias from etl.cfg, default is life """,
        default="life",
    )

    parser.add_argument(
        "-f", "--from_date", help="""from_date""", default=None
    )
    parser.add_argument("-t", "--to_date", help="""to_date""", default=None)
    parser.add_argument(
        "-b", "--batch_id", help="""enter batch id """, default=None
    )
    parser.add_argument(
        "-wf",
        "--batchy_job",
        help="""fully qualified batchy job name of the format wf.job""",
        default=None,
    )
    parser.add_argument(
        "-c",
        "--sql_command",
        help="""actual sql command to be executed""",
        default=None,
    )

    parser = ConfigWrapper.extend_parser(parser)

    args = parser.parse_args(args)

    (db_name, host, user, password, port, batchy_server, batchy_port) = ConfigWrapper.process_config(args)

    MSSQLRunner(
        db_name=db_name,
        host=host,
        user=user,
        password=password,
        port=port,
        batchy_server=batchy_server,
        batchy_port=batchy_port
    ).run_script(
        args.script,
        args.from_date,
        args.to_date,
        args.batch_id,
        args.parameters,
        args.batchy_job,
        args.sql_command,
    )


if __name__ == "__main__":
    main(sys.argv[1:])


mssql_runner/config_wrapper.py
from datacoco_core.config import config
from datacoco_secretsmanager import SecretsManager


class ConfigWrapper:
    """
    Wrapper file for config management for mssql_runner.
    """

    @staticmethod
    def sm_conf(project_name:str, team_name:str):
        """
        Simple config wrapper for using secrets manager.
        """
        c = SecretsManager().get_config(project_name, team_name)
        return c

    @staticmethod
    def extend_parser(parser):
        parser.add_argument(
            "-cfg",
            "--config",
            default="core",
            help="""
                whether to use secret_manager or
                datacoco_core to retrieve secrets
                """,
            choices=["secret_manager", "core"],
        )

        parser.add_argument(
            "-smp",
            "--secret_project_name",
            default="Maximilian3",
            help="""
                secret manager project group
                """
        )

        parser.add_argument(
            "-smt",
            "--secret_team",
            default="data",
            help="""
                secret manager team
                """
        )
        return parser

    @staticmethod
    def process_config(args):
        if args.config == "secret_manager":
            conf = ConfigWrapper.sm_conf(project_name=args.secret_project_name, team_name=args.secret_team)
        elif args.config == "core":
            conf = config()

        db_name = conf[args.database]["db_name"]
        host = conf[args.database]["host"]
        user = conf[args.database]["user"]
        password = conf[args.database]["password"]
        port = conf[args.database]["port"]
        batchy_server = conf["batchy"]["server"]
        batchy_port = conf["batchy"]["port"]

        return db_name, host, user, password, port, batchy_server, batchy_port


mssql_runner/mssql_runner.py
#!/usr/bin/env python

import sys
import os
import argparse

from datacoco_core.logger import Logger
from datacoco_batch import Batch
from datacoco_db.mssql_tools import MSSQLInteraction
from datetime import datetime
from datacoco_core.config import config

# import pymssql


# sys.excepthook = l.handle_exception
class MSSQLRunner:
    """
    generic class for execution of a parameterized script
    in postgres or redshift
    """

    def __init__(self, database):
        conf = config()
        self.conf = conf
        ms_db_name = conf[database]["db_name"]
        ms_user = conf[database]["user"]
        ms_server = conf[database]["server"]
        ms_password = conf[database]["password"]
        ms_port = conf[database].get("port", 1433)
        self.ms = MSSQLInteraction(
            dbname=ms_db_name,
            host=ms_server,
            user=ms_user,
            password=ms_password,
            port=ms_port,
        )
        self.ms.conn()

    @staticmethod
    def expand_params(sql, params):
        """
        substitutes params in sql stagement
        :param sql:
        :param params:
        :return: sql, expanded with params
        """
        for p in params.keys():
            var = "$[?" + p + "]"
            val = str(params[p])
            if p == "from_date" and val != "1776-07-04":
                try:
                    val = datetime.strptime(val, "%Y-%m-%dT%H:%M:%S.%f")
                    val = val.strftime("%Y-%m-%d %H:%M:%S")
                    print("removed microsecond from wf from_date")
                except Exception as e:
                    print(e)
                    pass
            sql = sql.replace(var, val)
        return sql

    def run_script(
        self,
        script,
        from_date=None,
        to_date=None,
        batch_id=None,
        params=None,
        batchy_job=None,
        sql_command=None,
    ):
        """
        method for expanding and running sql statements
        :param script:
        :param from_date:
        :param to_date:
        :param batch_id:
        :param params:
        :return:
        """

        paramset = {}

        # set up logger
        if script is not None:
            script_path, script_filename = os.path.split(script)
            logger = Logger(logname=script_filename)
        else:
            logger = Logger(logname="direct_sql")
            sys.excepthook = logger.handle_exception

        # first we retrive params  we will load these into dict first,
        # any additional params specified will override
        if batchy_job:
            wf = batchy_job.split(".")[0]
            try:
                job = (
                    batchy_job.split(".")[1]
                    if len(batchy_job.split(".")[1]) > 0
                    else "global"
                )
            except Exception as e:
                logger.l(e)
                job = "global"
            batchy_params = Batch(
                wf,
                server=self.conf["batchy"]["server"],
                port=self.conf["batchy"]["port"],
            ).get_status()
            paramset.update(batchy_params[job])

        # next we apply custom params and special metadata fields,
        # again this will overrite batchy params if specified
        # convert string params to dict
        try:
            params = dict(
                (k.strip(), v.strip())
                for k, v in (item.split("-") for item in params.split(","))
            )
        except Exception as e:
            logger.l(f"issue parsing params: {e}")

        if isinstance(params, dict):
            paramset.update(params)

        if from_date:
            paramset["from_date"] = from_date
        if to_date:
            paramset["to_date"] = to_date
        if batch_id:
            paramset["batch_id"] = batch_id

        # now defaults for special metadata fields
        if paramset.get("from_date") is None:
            paramset["from_date"] = "1776-07-04"
        if paramset.get("to_date") is None:
            paramset["to_date"] = "9999-12-31"
        if paramset.get("batch_id") is None:
            paramset["batch_id"] = "-1"
        # we'll keep batch_no for backwards compatibility
        paramset["batch_no"] = paramset["batch_id"]

        # check if it's direct SQL or SQL file
        if script is not None:
            try:
                raw_sql = open(script).read()
            except:
                e = f"File not found, please check file path."
                logger.l(e)
                raise RuntimeError(e)
        else:
            raw_sql = sql_command
        sql = self.expand_params(raw_sql, paramset)
        sql_message = "\n\n--sql script start:\n"+sql+"\n--sql script end\n\n"
        logger.l(sql_message)

        self.ms.batchOpen()

        logger.l("starting script")
        try:
            self.ms.exec_sql(sql)
            self.ms.batch_commit()
            logger.l("batch commit")
        except Exception as e:
            logger.l("execution failed with error: " + str(e))
            raise RuntimeError(e)
            # os._exit(1)  # this truncates logout in rundeck


if __name__ == "__main__":
    # argparse
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-s", "--script", help="""enter a path to your script """, default=None
    )
    # rundeck doesn't like -p '{"param1":"val1", "param2":"val2"}',
    # changed params to string
    parser.add_argument(
        "-p",
        "--parameters",  # type=json.loads, default='{"none":"none"}',
        default="none-none",
        help="""additional params to be substituted in script,
        example: -p param1-val1, param2-val2 """,
    )
    parser.add_argument(
        "-d",
        "--database",
        help="""db alias from etl.cfg, default is life """,
        default="life",
    )
    parser.add_argument(
        "-f",
        "--from_date",
        help="""from_date""",
        default=None
    )
    parser.add_argument(
        "-t",
        "--to_date",
        help="""to_date""",
        default=None
    )
    parser.add_argument(
        "-b",
        "--batch_id",
        help="""enter batch id """,
        default=None
    )
    parser.add_argument(
        "-wf",
        "--batchy_job",
        help="""fully qualified batchy job name of the format wf.job""",
        default=None,
    )
    parser.add_argument(
        "-c",
        "--sql_command",
        help="""actual sql command to be executed""",
        default=None,
    )
    args = parser.parse_args()

    MSSQLRunner(args.database).run_script(
        args.script,
        args.from_date,
        args.to_date,
        args.batch_id,
        args.parameters,
        args.batchy_job,
        args.sql_command,
    )
